### ISSUE TITLE
Improve the AppD's search function and documentation

### ISSUE THREAD

**Issue by kriswest** [finos/FDC3#370]
## Enhancement Request 
It is not currently clear from the documentation of the AppD search function what query parameters that it is intended support. The search function could/should support filtering of the app catalog by:
- name
- title
- description
- tags
- publisher
- keyword (a combination of title, description, tags, name, publisher - this could replace all other search filters above, apart from perhaps name & tags)

It should not support:
- search via AppId (as this is redundant with the Get call to retrieve an application definition `/v1/apps/{appId}`

In the documentation (https://fdc3.finos.org/schemas/1.2/app-directory#tag/Application/paths/~1v1~1apps~1search/get) it appears that the Query parameters area in the docs is being used to provide definitions of the fields returned for each application. The intended query parameters should be clarified and the field descriptions re-written to indicate how each field should be handled in the search.

### Use Case:
Searching for applications is a common use-case that should require no explanation.  Standardizing the search parameters will ensure that different UI implementations can correctly call the search functions of AppD's they were not developed or tested against. This is not guaranteed based on the current specification and documentation

### Current (confusing) documentation of query parameters
![image](https://user-images.githubusercontent.com/1701764/115709487-2efa9600-a369-11eb-81b3-18006550ab7d.png)

**Comment by thorsent** [comment]
The AppD API conflates two use cases into a single "search" API:

(1) Programmatic filtering - App Catalog interfaces wish to provide filtering based on various criteria.
(2) End user search - User's wish to search for applications that meet an arbitrary description.

Use of the word "search" for the first use case is misleading. I think that a single "GET list" interface that accepts narrowing criteria would be more precise (and which would return all apps by default).

The second use case would be better served by a single text based search field as end users'  have become accustomed toin modern applications. (The algorithm employed would be an implementation concern).

**Comment by robmoffat** [comment]
Can we have some examples of how this would look from a REST point of view?  i.e. how would these parameters be encoded?

**Comment by robmoffat** [comment]
A possibility is to recommend/standardize the use of GraphQL, outside the FDC3-specific intent-resolution APIs.

 Apart from anything else, this would mean we can modify the format of the appD record without worrying further about what the search API looked like

**Comment by kriswest** [comment]
Notes from meeting:
- #688

- There was generally consensus that the current search API parameters are not what most developers/users would expect and hence standardizing this API is not helping anyone.
  - `appId` parameter is redundant with the GET retrieval method
  - there are separate search parameters for text fields: `description`, `title`, `tooltip`
  - search by the `version` or manifest URL is unlikely to be useful
  - there are 3 separate fields for searching intents - usage is poorly described
- A proposal should be raised to update the definition of this API such that: 
  - a more conventional 'keyword' search is provided, where multiple text fields are searched including title, description, tooltip etc.
  - you can filter by category (with or without a keyword search)
  - search by intent name could/should be retained
    - however it was noted that this is severely limited by the inability to list all intents (and contexts) found in the app directory 
    - @kriswest@ we've seen a number of appD implementations add endpoints to list intents and contexts (including electron-fdc3 and the demo Citi provided)
      - An issue should be raised to add these endpoints to the standard to support filtering.
      - This also supports the use case for #247 (as contexts are regularly used through the broadcast API and intents may be raised rather than resolved, so the current config doesn't reference all intents and contexts that may be used)


@robmoffat 
> Can we have some examples of how this would look from a REST point of view? i.e. how would these parameters be encoded?

Yes next step would be to either describe a proposal in more detail in the issue (followed by a PR) or to raise a PR to do so.

> A possibility is to recommend/standardize the use of GraphQL, outside the FDC3-specific intent-resolution APIs.
>
> Apart from anything else, this would mean we can modify the format of the appD record without worrying further about what the search API looked like

Interesting suggestion, but does that actually solve a problem for Desktop Agents that will need to connect to different AppD implementations?  
- We don't need to retrieve field subsets (always return the whole record). 
- If we change the fields, then DA implementations will have to change the queries they use for search.
- Many of the fields in AppD are optional (details, title, tooltip, categories etc.) so they can expect variation in what's populated. 
  - The main application would be a keyword search over the whole set. AFAIK, to do that you need to construct a search field that combines them (rather than a complex query over multiple fields - which is better suited to retrieval than search). This ends up looking much like a keyword search that we could define as a standard REST endpoint.
- it's opinionated on the implementation: you have to incorporate GraphQL into the appD implementation and couldn't (easily) use another technology such as a search tool or database.

All that being said, I'm personally quite rusty on GraphML, hence, if you can flesh out a proposal further that might help me grok it.

Instead, I would recommend that we look at the most likely use cases (ways people will want to search/filter) and define a minimal REST interface that enables that. By keeping it minimal we limit changes needed due to new fields.

However, there is another alternative worth considering: don't have a search API in the AppD API at all. Use it purely as a repository of the records and have UIs retrieve them and then implement their own search procedures. This is feasible as the number of records is likely relatively small per appD and it allows the UIs to implement searches across a set of AppDs, rather than searching each and combining the results/presenting separate result sets...

**Comment by kriswest** [comment]
@robmoffat thanks for your comments, you've got me thinking about this differently.
@nkolba @ggeorgievx @openfin-johans @greyseer256 @Qiana-Citi interested in all your opinions on whether standardizing the AppD's search function really helps us at all?

The more I think about it, the more problems I see... Here are a few notes on use cases and implementation considerations which, I believe, indicate that we should not standardize/remove from the standard the search function:

## Use cases for searching AppD(s)
I believe the following represent the most likely use cases for searching app directories:

- In an app catalog or launcher UI a user wishes to:
  - Search for/filter applications via a standard keyword search box with queries such as:
    - "chart"
    - "Factset chart"
    - "OMS"
    - "position blotter"
    - "TCA"
  - Filter or browse applications by category, e.g.
    - "research"
    - "news"
    - "visualisation"
    - "communication"
  - Filter or browse applications by publisher
  - Find applications that work with a selected application (so that they can build up a workflow)
    - A subset of this use case would be searching for applications that can resolve a specific intent
    - The remainder of the use case being to search for applications that raise intents, apps that publish or listen to contexts on user channels, apps that publish or listen on particular named App Channels. 
      - See #247 

Other use cases that are possible but I think unlikely or far less common:
- A desktop agent wishes to search an appD programmatically as part of intent resolution
- A user wishes to search on specific text fields, rather than a keyword search over multiple text fields and the categories

However, we should acknowledge that as user interfaces for working with AppDs are not/should not be defined in the standard the list of use cases may be open ended.

### Implementation considerations
- An appD is a data source, a UI for browsing/searching will usually be a separate implementation from it,
  - Hence, a search function might be implemented client-side in the UI, server-side in the UI or (as is currently suggested) by query to the appD server.
- The dataset of all applications (across 1 or more than one appDs) is likely to be relatively small (compared to many search applications), say 10-500 records, hence:
  - a lookup by `title` or `publisher` is easy to perform via a table and sorting
  - the full dataset is trivial to retrieve in full to operate on locally
    - ideal scale for client-side search implementations...
    - whereas backing search with a technology such as ElasticSearch or Algolia would be overkill
- We wish to promote an ecosystem where vendors publish app directories in addition to appD's hosted in-house by organisations (e.g banks) consuming apps
  - Searching multiple directories would involve submitting queries to each directory and presenting combined results
    - Different appD implementations might handle search implementation differently (whole string match, substring match, with or without keyword stemming, weighted (TF/IDF) keyword etc.) - difficult to know how to combine results into a single list
    - To build filtering interfaces, each directory would have to be queried for the full set of publishers, categories, intents etc. that are referenced within it and those sets combined.

## How does standardization of search functions help?
For apps/UIs that consume AppD:
- If you are using only a single AppD, then you can rely on the AppD for your search implementation 👍 
  - sets a minimum requirement (for standard compliance) on the available search functions 👍 
  - but also constrains the search functions you can implement 👎 
- but if using multiple then you have an aggregation problem due to potential differences in the behaviour of different appDs 👎 
  - Only resolvable by defining how searches should be implemented 👎 

For vendors and implementors of AppDs:
- Adds technical complexity to the implementation of the AppD 👎
- May increase hosting requirements (without a search, entitlement or create function an appD could actually be implemented as static documents at the required paths...) 

To summarize, if I were to implement an interface over multiple AppDs, I wouldn't use the search functions in the current or any future API. Instead, I would retrieve all the records (server or client-side) and implement search in the UI itself - probably with a JS lib such as [js-search](https://github.com/bvaughn/js-search), [Fuse](https://fusejs.io/), [Lunr](https://lunrjs.com/docs/index.html) or any number of autocomplete libraries.

Further, if I'm working with a single in-house AppD and in-house launcher UI, applying a standard to this function doesn't gain me anything and might actually be constraining...
